python/lsst/daf/butler/gen2convert/writer.py

```diff
@@ -0,0 +1,316 @@
+                         "with name={}".format(skyMapName, sha1.hex(), existing["skymap"]))
+                    )
+                continue
+            except ValueError:
+                # No SkyMap with this sha1 exists, so we need to insert it.
+                pass
+            if skyMapName is None:
+                raise LookupError(
+                    ("SkyMap with sha1={} has no name "
+                     "and does not already exist in the Registry.").format(sha1.hex())
+                )
+            log.info("Inserting SkyMap '%s' with sha1=%s", skyMapName, sha1.hex())
+            skyMap.register(skyMapName, registry)
+
+    def insertObservations(self, registry):
+        """Add all necessary Visit and Exposure DataUnits to the Registry.
+        """
+        log = Log.getLogger("lsst.daf.butler.gen2convert")
+        for mapperName, nested in self.visitInfo.items():
+            camera = self.config["mappers"][mapperName]["camera"]
+            log.info("Inserting Exposure and Visit DataUnits for Camera '%s'", camera)
+            for visitInfoId, (visitInfo, filt) in nested.items():
+                # TODO: generalize this to cameras with snaps and/or compound gen2 visit/exposure IDs
+                visitId, = visitInfoId
+                exposureId, = visitInfoId
+                # TODO: skip insertion if DataUnits already exist.
+                mid = visitInfo.getDate().toPython()
+                offset = datetime.timedelta(seconds=0.5*visitInfo.getExposureTime())
+                commonValues = {
+                    "camera": camera,
+                    "visit": visitId,
+                    "physical_filter": filt,
+                    "datetime_begin": mid - offset,
+                    "exposure_time": visitInfo.getExposureTime(),
+                    "boresight_az": visitInfo.getBoresightAzAlt().getLongitude().asDegrees(),
+                    "boresight_alt": visitInfo.getBoresightAzAlt().getLatitude().asDegrees(),
+                    "rot_angle": visitInfo.getBoresightRotAngle().asDegrees(),
+
+                }
+                exposureValues = commonValues.copy()
+                exposureValues.update({
+                    "exposure": exposureId,
+                    "snap": 0,
+                    "dark_time": visitInfo.getDarkTime()
+                })
+                visitValues = commonValues.copy()
+                visitValues.update({
+                    "datetime_end": mid + offset,
+                    "earth_rotation_angle": visitInfo.getEra().asDegrees(),
+                    "boresight_ra": visitInfo.getBoresightRaDec().getLongitude().asDegrees(),
+                    "boresight_dec": visitInfo.getBoresightRaDec().getLatitude().asDegrees(),
+                    "boresight_parallactic_angle": visitInfo.getBoresightParAngle().asDegrees(),
+                    "local_era": visitInfo.getLocalEra().asDegrees(),
```

Comment:
Do we loose information with respect to `visitInfo` here?

Reply:
Possibly, but I don't want to think too hard about this until we've gone through the schema review.